### DISCUSSION TITLE
Handle event in javascript Livewire directive

### DISCUSSION THREAD

**Discussion by svenbw** [livewire/livewire#5035] [1. Help]
I created a directive using `window.Livewire.directive`, an example usage would look in the Livewire component like this:
```html
<div>
    <!-- component html -->

    <div wire:myDirective>
        <!-- directive html -->
    </div>

    <!-- component html -->
</div>
```
If one needs to send a event to a specific component one typically uses `$this->emitTo('MyComponentUsingMyDirective', 'eventName', 1234)`.

The only way I could catch these events for my directive within within that Livewire component is by hooking to all messages and filtering based on the name of the component and check if the event that should be handled by the directive itself matches the name of the event (_eventName_ in the example below)
Any arguments passed with the arguments can be extracted from `event.params` (`argumentOne` will be _1234_)

```JavaScript
window.Livewire.directive('sortable', (el, directive, component) => {

    // ... code to create the directive

    window.Livewire.hook('message.received', (message) => {
        const effects = message.response.effects
        if (effects.emits && effects.emits.length > 0) {
            effects.emits.forEach(event => {
                if (event.to === component.fingerprint.name) {
                    if (event.event === 'eventName') {
                        const argumentOne = event.params[0]
                        // ... do something with the directive component if eventName is fired
                    }
                }
            })
        }
    });
})
```

I couldn't find any other solution to this, if anybody has any suggestions...

**Comment by joshhanley** [comment]
@svenbw can you not just use a Livewire event listener `Livewire.on()`? https://laravel-livewire.com/docs/2.x/events#in-js

**Reply by svenbw** [reply]
The `Livewire.on()` event listener listens to events without a scope.

This means that if you would use the same directive in 2 child components both directives would react to the same event, although you want to target only one. Therefore I need some kind of filtering so I can listen to scoped event in Javascript.

**Reply by joshhanley** [reply]
@svenbw ah right ok.

Is the code in the example above your actual code? Noticed you have `wire:myDirective` in the html and `sortable` in the JS, which doesn't match up. Can you share your actual directive, use case/ what you're intending to do and code so I can see how you're intending it to work and maybe we can come up with a better way 🙂

I've got a few custom directives in one of my projects I can look and up compare too.

**Reply by svenbw** [reply]
@joshhanley more or less, it's a simplified version of it, sharing all the code would clutter all the question (or at least I thought so).

In the example below I will try to explain the issue based on a directive that can interact with full-calendar. I stripped down additional markup for this example. Also both calendar components are equal, but again this is to keep the clutter from this question. Also the directive itself is stripped down to the bare minimum.

**blade view** that holds the wrapper Livewire component
```blade
@section('page')
    <div class="row row-deck row-cards">
        <div class="col-12">
            <livewire:demo.wrapper />
        </div>
    </div>
@endsection
```

**Livewire wrapper component** this component will receive the clicks from two buttons which move the calendar to the next month:
```php
use Livewire\Component;

class Wrapper extends Component
{
    public function nextMonth(int $calendar)
    {
        if ($calendar === 1) {
            // $this->emit('CalendarNext'); // <- targets Livewire.on()
            $this->emitTo('demo.calendar1', 'CalendarNextTo'); // <- new scoped emit
        } else if ($calendar === 2) {
            $this->emitTo('demo.calendar2', 'CalendarNextTo');
            // $this->emit('CalendarNext');
        }
    }

    public function render()
    {
        return view('livewire.demo.wrapper');
    }
}
```
```blade
<div>
    <button wire:click="nextMonth(1)" class="btn btn-primary">Next month 1</button>
    <button wire:click="nextMonth(2)" class="btn btn-primary">Next month 2</button>

    <livewire:demo.calendar1 />
    <livewire:demo.calendar2 />
</div>
```

**Livewire calendar component** this holds the `wire:calendar` directive (note that: the classes themselves just contain the render function, and the 1 will be replaced with 2 for the second component)
```blade
<div>
    <h3>Calendar 1</h3>
    <div
        id="calendar1"
        wire:calendar
    ></div>
</div>
```

Finally the directive that wraps the full calendar component.
```js
if (window.Livewire) {
    window.Livewire.directive('calendar', (el, directive, component) => {
        el.livewire_full_calendar = new Calendar(el, {
            plugins: [
                interactionPlugin,
                dayGridPlugin,
                timeGridPlugin,
                listPlugin,
            ],
        });

        el.livewire_full_calendar.render();

        window.Livewire.on('CalendarNext', () => {
            el.livewire_full_calendar.next()
        })

        window.Livewire.on('CalendarPrev', () => {
            el.livewire_full_calendar.prev()
        })

        window.Livewire.hook('message.received', (message) => {
            const effects = message.response.effects
            if (effects.emits && effects.emits.length > 0) {
                effects.emits.forEach(event => {
                    if (event.to === component.fingerprint.name) {
                        if (event.event === 'CalendarNextTo') {
                            el.livewire_full_calendar.next()
                        }
                        else if (event.event === 'CalendarPrevTo') {
                            el.livewire_full_calendar.prev()
                        }
                    }
                })
            }
        });
    });
}
```
As mentioned above the directive is stripped down to the initialization and event handling, in the full directive I'm able to pass the event from the Livewire component class to the calender, attach event handlers so I can handle selections/edit actions directly in the Livewire component. So a great match.

Thanks for looking into this!

**Reply by joshhanley** [reply]
@svenbw ah yep makes sense. I will have a look sometime next week and see how I'm handling it and if there is a way we can simplify your code 🙂 but if it works, I'd also say there's nothing wrong with it 😁
